### PR TITLE
[WIP] Dev

### PR DESCRIPTION
**Pull request by erogol**
- bidirectional decoding (https://arxiv.org/abs/1907.09006), it provides more robust alignment, especially for hard datasets.
- [WIP] some code refactoring
- continuing training from where it is left off without creating a new folder.
- Speaker encoding model as a subproject following https://arxiv.org/abs/1806.04558
- Release a pre-trained Speaker Encoder trained on LibriTTS.
- Graves Attention (It is more stable for some datasets) (thx @geneing )
- [WIP] Multi-Head Attention
- [WIP] ParallelWaveGAN integration https://github.com/erogol/ParallelWaveGAN
   - it provides faster than real-time GPU inference with TTS and almost real-time CPU inference.
- Speaker embedding visualization by @nmstoker 
- Easier server usage by @reuben 
- Fix memory overload for restoring a model.
- [WIP] Fastspeech https://arxiv.org/abs/1905.09263
- [WIP] Alternative Tacotron architectures.
- [WIP] Training with gradient accumulation for small batch sizes. 
- TestAttention notebook for testing model attention performance on 50 hard sentences.

### CODE REVIEW

File: layers/common_layers.py [hunk 1]
```diff
@@ -105,14 +105,100 @@ def forward(self, attention_cat):
...
+        self.mu_prev = None
+        self.init_layers()
+
+    def init_layers(self):
+        torch.nn.init.constant_(self.N_a[2].bias[10:15], 0.5)
+        torch.nn.init.constant_(self.N_a[2].bias[5:10], 10)
+
+    def init_states(self, inputs):
+        if self.J is None or inputs.shape[1] > self.J.shape[-1]:
+            self.J = torch.arange(0, inputs.shape[1]).to(inputs.device).expand([inputs.shape[0], self.K, inputs.shape[1]])
```
Comment (tux3) [inline layers/common_layers.py]: I tried out this branch out of curiosity, but I had a dtype mismatch in `forward` due to `J` being a Long. I'm really not familiar with the code, but would `torch.arange(0.0` be a reasonable fix?

Comment (erogol) [reply layers/common_layers.py]: which version pytorch?

Comment (tux3) [reply layers/common_layers.py]: Pytorch 1.1.0 (on python 3.7.5)

Comment (erogol) [reply layers/common_layers.py]: try the latest pytorch, I guess the behaviour has been changed. But yes you can also solve as you said.
